Create a type safe DSL for edges

karloti opened the pull request:
### The 'edge' function takes a lambda as an argument.
```
edge {
    nodeStart forwardTo nodeCallLLM
}
```
---

#### Type of the change
- [x] New feature
- [ ] Bug fix
- [ ] Documentation fix

#### Checklist for all pull requests
- [x] The pull request has a description of the proposed change
- [x] I read the [Contributing Guidelines](https://github.com/JetBrains/koog/blob/main/CONTRIBUTING.md) before opening the pull request
- [x] The pull request uses **`develop`** as the base branch
- [x] Tests for the changes have been added
- [x] All new and existing tests passed

##### Additional steps for pull requests adding a new feature
- [ ] An issue describing the proposed change exists
- [ ] The pull request includes a link to the issue
- [ ] The change was discussed and approved in the issue
- [ ] Docs have been added / updated

ptitjes commented:
Maybe one mitigation would be to add a more involved graph verification. IRC, we currently only check for a path from start to finish.

karloti commented:
```
val strategy = strategy("test-interceptors-strategy") {
            val llmCallWithoutTools by nodeLLMRequest("test LLM call", allowToolCalls = false)
            val llmCall by nodeLLMRequest("test LLM call with tools")

            nodeStart forwardTo llmCallWithoutTools transformed { "Test LLM call prompt" }
            llmCallWithoutTools forwardTo llmCall transformed { "Test LLM call with tools prompt" }
            llmCall forwardTo nodeFinish transformed { "Done" }
        }

```

Add tests and update edge handling in AIAgentNode

- Introduced unit tests for AIAgentNode to validate edge additions and replacements.
- Refactored `edges` to use a mutable map for managing connections, allowing edge replacement functionality.
- Enhanced AIAgentEdgeBuilder and pipeline interceptors for better event handling and functionality testing.

Ololoshechkin commented:
Dear @karloti, thanks for your API proposal and MR — it looks great and has some really cool ideas!

I've reviewed the implementation and experimented with a few strategies. One concern I noticed is that the current approach seems to result in a loss of type-safety. Here's an example:
```kotlin
val strategy = strategy("test") {
    val hash by node<String, Int> { it.hashCode() }
    val add5 by node<Int, Int> { it + 5 }
    val presentResult by node<String, String> { "result: $it" }

    nodeStart forwardTo hash
    hash forwardTo add5
    add5 forwardTo presentResult // `edge(add5 forwardTo presentResult)` -- would've been a type error with the original API
    presentResult forwardTo nodeFinish
}
```

And the following full example will crash in runtime :
```
@Test
    fun testF() = runTest{
        val strategy = strategy("test") {
            val hash by node<String, Int> { it.hashCode() }
            val add5 by node<Int, Int> { it + 5 }
            val presentResult by node<String, String> { "result: $it" }

            nodeStart forwardTo hash
            hash forwardTo add5
            add5 forwardTo presentResult // edge(add5 forwardTo presentResult) -- type error with the original API
            presentResult forwardTo nodeFinish
        }

        // Use custom tool registry with plus tool to be called
        val toolRegistry = ToolRegistry {
            tool(CalculatorTools.PlusTool)
        }

        val agent = createAgent(
            strategy = strategy,
            toolRegistry = toolRegistry,
            userPrompt = "add 2.2 and 2.2",
            promptExecutor = CalculatorChatExecutor
        )

        val result = agent.runAndGetResult("Hello World!")

        println(result)
    }
    ```
    
    The error in runtime will be:
    ```
    class java.lang.Integer cannot be cast to class java.lang.String (java.lang.Integer and java.lang.String are in module java.base of loader 'bootstrap')
java.lang.ClassCastException: class java.lang.Integer cannot be cast to class java.lang.String (java.lang.Integer and java.lang.String are in module java.base of loader 'bootstrap')
	at ai.koog.agents.core.feature.AIAgentPipelineTestV2$testF$1$strategy$1$presentResult$2.invoke(AIAgentPipelineTestV2.kt)
	at ai.koog.agents.core.agent.entity.AIAgentNode.execute(AIAgentNode.kt:139)
	at ai.koog.agents.core.agent.entity.AIAgentNodeBase.executeUnsafe(AIAgentNode.kt:118)
	at ai.koog.agents.core.agent.entity.AIAgentSubgraph.doExecute(AIAgentSubgraph.kt:78)
	...
```
